File: src/posidon/uranium/engine/Input.java
package posidon.uranium.engine;

import org.lwjgl.glfw.*;

public class Input {
    private static boolean[] keys = new boolean[GLFW.GLFW_KEY_LAST];
    private static boolean[] mouseButtons = new boolean[GLFW.GLFW_MOUSE_BUTTON_LAST];
    private static double curX, curY;
    private static double scrollX, scrollY;

    private GLFWKeyCallback keyListener;
    private GLFWCursorPosCallback cursorListener;
    private GLFWMouseButtonCallback mouseButtonListener;
    private GLFWScrollCallback scrollListener;

    public Input(Window w) {
        keyListener = new GLFWKeyCallback() {
            @Override
            public void invoke(long window, int key, int scanCode, int action, int mods) {
                keys[key] = (action != GLFW.GLFW_RELEASE);
                if (key == GLFW.GLFW_KEY_F11) w.setFullscreen(!w.isFullscreen());
                else if (key == GLFW.GLFW_KEY_ESCAPE) w.setMouseLocked(false);
            }
        };

        cursorListener = new GLFWCursorPosCallback() {
            @Override
            public void invoke(long window, double x, double y) {
                if (Window.mouseLocked) {
                    curX = x;
                    curY = y;
                }
            }
        };

        mouseButtonListener = new GLFWMouseButtonCallback() {
            @Override
            public void invoke(long window, int btn, int action, int mods) {
                mouseButtons[btn] = (action != GLFW.GLFW_RELEASE);
                if (btn == GLFW.GLFW_MOUSE_BUTTON_LEFT) w.setMouseLocked(true);
            }
        };

        scrollListener = new GLFWScrollCallback() {
            @Override
            public void invoke(long window, double x, double y) {
                scrollX += x;
                scrollY += y;
            }
        };
    }

    public static boolean isKeyDown(int key) { return keys[key]; }
    public static boolean isButtonDown(int btn) { return mouseButtons[btn]; }

    public void kill() {
        keyListener.free();
        cursorListener.free();
        mouseButtonListener.free();
    }

    public static double getCurX() { return curX; }
    public static double getCurY() { return curY; }
    public static double getScrollX() { return scrollX; }
    public static double getScrollY() { return scrollY; }
    public GLFWKeyCallback getKeyListener() { return keyListener; }
    public GLFWCursorPosCallback getCursorListener() { return cursorListener; }
    public GLFWMouseButtonCallback getMouseButtonListener() { return mouseButtonListener; }
    public GLFWScrollCallback getScrollListener() { return scrollListener; }
}

File: src/posidon/uranium/engine/Input.kt
package posidon.uranium.engine

import org.lwjgl.glfw.*

class Input(w: Window) {

    val keyListener: GLFWKeyCallback
    val cursorListener: GLFWCursorPosCallback
    val mouseButtonListener: GLFWMouseButtonCallback
    val scrollListener: GLFWScrollCallback

    fun kill() {
        keyListener.free()
        cursorListener.free()
        mouseButtonListener.free()
        scrollListener.free()
    }

    companion object {
        private val keys = BooleanArray(GLFW.GLFW_KEY_LAST)
        private val mouseButtons = BooleanArray(GLFW.GLFW_MOUSE_BUTTON_LAST)
        var curX = 0.0
        var curY = 0.0
        var scrollX = 0.0
        var scrollY = 0.0
        fun isKeyDown(key: Int): Boolean = keys[key]
        fun isButtonDown(btn: Int): Boolean = mouseButtons[btn]
    }

    init {
        keyListener = object : GLFWKeyCallback() {
            override fun invoke(window: Long, key: Int, scanCode: Int, action: Int, mods: Int) {
                try {
                    keys[key] = action != GLFW.GLFW_RELEASE
                    println(action)
                    when (key) {
                        GLFW.GLFW_KEY_F11 -> w.isFullscreen = !w.isFullscreen
                        GLFW.GLFW_KEY_ESCAPE -> w.setMouseLocked(false)
                    }
                } catch (e: Exception) { e.printStackTrace() }
            }
        }
        cursorListener = object : GLFWCursorPosCallback() {
            override fun invoke(window: Long, x: Double, y: Double) {
                if (Window.mouseLocked) {
                    curX = x
                    curY = y
                }
            }
        }
        mouseButtonListener = object : GLFWMouseButtonCallback() {
            override fun invoke(window: Long, btn: Int, action: Int, mods: Int) {
                mouseButtons[btn] = action != GLFW.GLFW_RELEASE
                if (btn == GLFW.GLFW_MOUSE_BUTTON_LEFT) w.setMouseLocked(true)
            }
        }
        scrollListener = object : GLFWScrollCallback() {
            override fun invoke(window: Long, x: Double, y: Double) {
                scrollX += x
                scrollY += y
            }
        }
    }
}